openshift/machine-config-operator
# MCO-1029: MCO-678: Implement Boot Image update opt-in API

### This PR does the following:

- The MSBIC now checks/watches the MachineConfiguration knob for boot image management settings and runs a sync loop of the enrolled machinesets
- If any of the enrolled machinesets have an OwnerReference, the MSBIC will error & report the misbehaving machineset. An OwnerReference implies the machineset is managed by another workflow, and should not be updated by the MCO.
- The MSBIC will now point all machinesets to the managed secret stub `worker-user-data-managed`
- The e2e-tech-preview test has been updated to test different variations of the boot image management configuration knob

### How to test:

With this PR, to turn on boot image management, we need to:

- Turn on TechPreview featureset
- Update the MachineConfiguration Knob to opt in some or all machinesets. To do so, apply the following CR:

To select all machinesets:
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
  logLevel: Normal
  operatorLogLevel: Normal
  managedBootImages:
    machineManagers:
    - resource: machinesets
      apiGroup: machine.openshift.io
      selection:
        mode: All
```
To partially select some machinesets, use the label selector in partial mode. Apply the label `test: boot` on the target machineset(s) and apply the following object. The order of applying should not matter.
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
  logLevel: Normal
  operatorLogLevel: Normal
  managedBootImages:
    machineManagers:
    - resource: machinesets
      apiGroup: machine.openshift.io
      selection:
        mode: Partial
        partial:
          machineResourceSelector:
            matchLabels:
              test: boot
```
Please note that the above objects will be rejected on any non-GCP cluster as https://github.com/openshift/machine-config-operator/pull/4285 has landed, so be sure you are testing on GCP. Creating the above object should automatically initiate another sync loop and reconcile any enrolled machinesets. If this object does not exist, no machinesets should be reconciled. 
- On any enrolled machinesets, ensure that the stub secret being used is the managed version by checking the machineset's providerspec. 
- If any enrolled machineset has an ownerreference, this should now cause the MSBIC to error out and eventually cause a Prometheus alert. This will be updated to a condition on the MachineConfiguration object prior to GA.
- As for the e2e tests, passing e2e-gcp-op-techpreview should be enough to validate that piece of work.
